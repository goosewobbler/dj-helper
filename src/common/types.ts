import { AnyAction } from 'redux';
import { ThunkDispatch } from 'redux-thunk';

enum ModuleType {
  View,
  ViewCSS,
  Data,
}

enum ComponentType {
  Page = 1,
  View,
  Data,
}

enum ComponentState {
  Stopped = 1,
  Starting,
  Installing,
  Building,
  Running,
  Linking,
}

interface AppState {
  components: ComponentData[];
  ui: {
    componentToClone?: string;
    editors: string[];
    selectedComponent?: string;
    filter?: string;
    updating?: boolean;
    updated?: boolean;
    showDialog?: string;
    hideDialog?: boolean;
  };
}

interface Component {
  bump(type: BumpType): Promise<void>;
  fetchDetails(): Promise<void>;
  getName(): string;
  getDirectoryName(): string;
  getDisplayName(): string;
  getType(): ComponentType;
  getURL(): string | null;
  getFavourite(): boolean;
  getHistory(): string[];
  getUseCache(): boolean;
  setFavourite(favourite: boolean): Promise<void>;
  setUseCache(useCache: boolean): Promise<void>;
  getDependencies(): ComponentDependency[];
  getDependenciesSummary(): Promise<{ name: string }[]>;
  getLatestVersion(): Promise<string>;
  getLinking(): string[];
  getVersions(): Versions;
  getState(): ComponentState;
  getPromoting(): string | null;
  getPromotionFailure(): string | null;
  getRendererType(): string;
  promote(environment: string): Promise<void>;
  openInEditor(): Promise<void>;
  reinstall(): Promise<void>;
  link(dependency: string): Promise<void>;
  unlink(dependency: string): Promise<void>;
  makeLinkable(): Promise<void>;
  build(isSassOnly?: boolean, path?: string): Promise<void>;
  setPagePort(pagePort: number): void;
  start(): Promise<void>;
  stop(): Promise<void>;
  request(props: LooseObject, history: boolean): Promise<Response>;
}

interface ComponentsData {
  components: ComponentData[];
  editors: string[];
}

enum BumpType {
  patch = 'patch',
  minor = 'minor',
}

type ComponentMatch = string | { matched: string };

interface ComponentData {
  name: string;
  displayName: string;
  highlighted?: React.ReactElement[];
  state: ComponentState;
  favourite: boolean;
  history?: string[];
  url?: string;
  type?: ComponentType;
  dependencies?: ComponentDependency[];
  linking?: string[];
  promoting?: string;
  promotionFailure?: string;
  useCache: boolean;
  versions?: Versions;
  rendererType: string;
  alternatives?: string[];
  matches?: ComponentMatch[];
}

interface ComponentDependency {
  name: string;
  displayName: string;
  has: string | null;
  latest: string | null;
  linked: boolean;
  outdated: boolean;
  version: string | null;
  rendererType: string | null;
}

type GraphNode = {
  id: number;
  name: string;
};

type GraphData = {
  edges: {
    from: number;
    to: number;
  }[];
  nodes: GraphNode[];
};

interface Response {
  body: string;
  headers: Headers;
  statusCode: number;
}

type StringObject = { [Key: string]: string };

interface Shrinkwrap {
  dependencies: {
    [Key: string]: StringObject;
  };
}

interface Package {
  name: string;
  description: string;
  homepage: string;
  scripts: StringObject;
  dependencies: StringObject;
  devDependencies: StringObject;
  version: string;
  morph: { rendererType: string };
}

type AnyObject = Record<string, unknown>;

interface LooseObject {
  [Key: string]: AnyObject | [] | string | number | boolean;
}

interface AppStatus {
  currentVersion: string;
  updateAvailable: string;
  updated: boolean;
  updating: boolean;
}

interface FileSystem {
  exists(path: string): Promise<boolean>;
  getPackageDirectories(directory: string): string[];
  readFile(path: string): Promise<string>;
  writeFile(path: string, contents: string): Promise<void>;
  symbolicLinkExists(path: string): Promise<boolean>;
  copyDirectory(from: string, to: string, filter: boolean): Promise<void>;
  deleteDirectory(directory: string): void;
  watchDirectory(directory: string, callback: (path: string) => void): Promise<void>;
  moveDirectory(from: string, to: string): void;
  createSymlink(from: string, to: string): void;
  removeSymlink(path: string): void;
}

interface GitSystem {
  checkoutMaster(directory: string): Promise<void>;
  checkoutExistingBranch(directory: string, branchName: string): Promise<void>;
  checkoutNewBranch(directory: string, branchName: string): Promise<void>;
  commit(directory: string, message: string): Promise<void>;
  getCurrentBranch(directory: string): Promise<string>;
  getRandomBranchName(): Promise<string>;
  push(directory: string, branchName: string): Promise<void>;
  readyToCommit(directory: string): Promise<boolean>;
  stageFile(directory: string, path: string): Promise<void>;
}

interface MorphSystem {
  getShrinkwrap(name: string): Promise<{ [Key: string]: string }>;
  getVersionOnEnvironment(name: string, environment: string): Promise<string>;
  promote(name: string, environment: string): Promise<void>;
}

interface NetworkSystem {
  get(url: string): Promise<Response | undefined>;
}

interface ProcessSystem {
  getCurrentWorkingDirectory(): Promise<string>;
  log(message: string): void;
  open(url: string): void;
  runToCompletion(
    directory: string,
    command: string,
    onOutput: (message: string) => void,
    onError: (message: string) => void,
  ): Promise<void>;
  runUntilStopped(
    directory: string,
    command: string,
    onOutput: (message: string) => void,
    onError: (message: string) => void,
  ): Promise<() => Promise<void>>;
}

interface System {
  file: FileSystem;
  git: GitSystem;
  morph: MorphSystem;
  network: NetworkSystem;
  process: ProcessSystem;
}

type Dispatch = ThunkDispatch<AppState, undefined, AnyAction>;

type StoredValue = string[] | string | number | boolean;

interface ValueStore {
  [Key: string]: StoredValue;
}

interface Store {
  get(key: string): StoredValue | null;
  set(key: string, value: StoredValue | null): Promise<void>;
}

// type EnvironmentValues = 'int' | 'local' | 'live' | 'test';
// enum EnvironmentValues {
//   int = 'int',
//   local = 'local',
//   live = 'live',
//   test = 'test',
// }

type Versions = {
  [key: string]: string | null;
};
// type Versions = { [key in EnvironmentValues]: string | null };

interface Service {
  bump(name: string, type: BumpType): Promise<void>;
  build(name: string): Promise<void>;
  clone(name: string, cloneName: string, options: { description: string }): Promise<void>;
  create(name: string, type: ModuleType, options: { description: string }): Promise<void>;
  fetchDetails(name: string): Promise<void>;
  getComponentsData(): ComponentsData;
  getComponentsSummaryData(): ComponentsData;
  getDependantGraph(name: string): GraphData;
  getDependencyGraph(name: string): GraphData;
  link(name: string, dependency: string): Promise<void>;
  openInEditor(name: string): Promise<void>;
  promote(name: string, environment: string): Promise<void>;
  reinstall(name: string): Promise<void>;
  request(name: string, props: LooseObject, history: boolean): Promise<Response>;
  setFavourite(name: string, favourite: boolean): Promise<void>;
  setUseCache(name: string, useCache: boolean): Promise<void>;
  start(name: string): Promise<void>;
  stop(name: string): Promise<void>;
  unlink(name: string, dependency: string): Promise<void>;
}

export type ComponentHandlers = {
  buildComponent(name: string): void;
  bumpComponent(name: string, type: string): void;
  showCloneComponentDialog(sourceComponent: string): void;
  installComponent(name: string): void;
  linkComponent(name: string, dependency: string): void;
  promoteComponent(name: string, dependency: string): void;
  updateAndSelectComponent(name: string): void;
  setUseCacheOnComponent(name: string, value: boolean): void;
  unlinkComponent(name: string, dependency: string): void;
  openInCode(name: string): void;
};

export {
  AnyObject,
  Shrinkwrap,
  StringObject,
  Versions,
  LooseObject,
  BumpType,
  Component,
  ComponentData,
  ComponentDependency,
  ComponentMatch,
  ComponentState,
  ComponentType,
  ComponentsData,
  GraphData,
  GraphNode,
  ModuleType,
  Response,
  Package,
  Service,
  AppStatus,
  AppState,
  Dispatch,
  System,
  FileSystem,
  GitSystem,
  MorphSystem,
  NetworkSystem,
  ProcessSystem,
  Store,
  StoredValue,
  ValueStore,
};
