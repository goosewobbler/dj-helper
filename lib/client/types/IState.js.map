{"version":3,"file":"IState.js","sourceRoot":"","sources":["../../../src/client/types/IState.ts"],"names":[],"mappings":""}