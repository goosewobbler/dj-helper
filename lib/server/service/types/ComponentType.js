"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var ComponentType;
(function (ComponentType) {
    ComponentType[ComponentType["Page"] = 1] = "Page";
    ComponentType[ComponentType["View"] = 2] = "View";
    ComponentType[ComponentType["Data"] = 3] = "Data";
})(ComponentType || (ComponentType = {}));
exports.default = ComponentType;
//# sourceMappingURL=ComponentType.js.map