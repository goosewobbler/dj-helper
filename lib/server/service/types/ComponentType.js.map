{"version":3,"file":"ComponentType.js","sourceRoot":"","sources":["../../../../src/server/service/types/ComponentType.ts"],"names":[],"mappings":";;AAAA,IAAK,aAIJ;AAJD,WAAK,aAAa;IAChB,iDAAQ,CAAA;IACR,iDAAI,CAAA;IACJ,iDAAI,CAAA;AACN,CAAC,EAJI,aAAa,KAAb,aAAa,QAIjB;AAED,kBAAe,aAAa,CAAC"}