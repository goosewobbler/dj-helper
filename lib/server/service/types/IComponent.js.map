{"version":3,"file":"IComponent.js","sourceRoot":"","sources":["../../../../src/server/service/types/IComponent.ts"],"names":[],"mappings":""}