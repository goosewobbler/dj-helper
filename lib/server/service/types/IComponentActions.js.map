{"version":3,"file":"IComponentActions.js","sourceRoot":"","sources":["../../../../src/server/service/types/IComponentActions.ts"],"names":[],"mappings":""}