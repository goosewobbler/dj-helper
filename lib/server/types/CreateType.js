"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var CreateType;
(function (CreateType) {
    CreateType[CreateType["View"] = 0] = "View";
    CreateType[CreateType["ViewCSS"] = 1] = "ViewCSS";
    CreateType[CreateType["Data"] = 2] = "Data";
})(CreateType || (CreateType = {}));
exports.default = CreateType;
//# sourceMappingURL=CreateType.js.map