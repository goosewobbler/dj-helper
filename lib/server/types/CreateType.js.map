{"version":3,"file":"CreateType.js","sourceRoot":"","sources":["../../../src/server/types/CreateType.ts"],"names":[],"mappings":";;AAAA,IAAK,UAIJ;AAJD,WAAK,UAAU;IACb,2CAAI,CAAA;IACJ,iDAAO,CAAA;IACP,2CAAI,CAAA;AACN,CAAC,EAJI,UAAU,KAAV,UAAU,QAId;AAED,kBAAe,UAAU,CAAC"}