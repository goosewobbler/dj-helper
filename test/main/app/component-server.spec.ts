import * as request from 'supertest';
import express from 'express';
import createServer from '../../../src/main/app/componentServer';
import createMockService from '../mocks/service';
import mockConfig from '../mocks/config';
import { Service } from '../../../src/common/types';

let server: express.Express;
let mockService: Service;

const mockServiceRequest = (body: string, headers: Record<string, string>, statusCode: number): void => {
  (mockService.request as jest.Mock).mockReturnValueOnce(
    Promise.resolve({
      body,
      headers,
      statusCode,
    }),
  );
};

beforeEach(() => {
  mockService = createMockService();
  server = createServer(mockService, mockConfig);
});

afterEach(() => {
  (mockService.request as jest.Mock).mockReset();
});

describe('when a given component is not running', () => {
  beforeEach(() => {
    mockServiceRequest('Component is not running', {}, 500);
  });

  it('cannot request a data component', (done) =>
    request(server)
      .get('/data/bbc-morph-baz')
      .expect('Content-Type', 'text/html; charset=utf-8')
      .expect(500)
      .then((response: request.Response): void => {
        expect(mockService.request).toHaveBeenCalledWith('bbc-morph-baz', {}, true);
        expect(response.text).toEqual('Component is not running');
      })
      .then(done));

  it('cannot request a view component', (done) =>
    request(server)
      .get('/view/bbc-morph-baz')
      .expect('Content-Type', 'text/html; charset=utf-8')
      .expect(500)
      .then((response: request.Response): void => {
        expect(mockService.request).toHaveBeenCalledWith('bbc-morph-baz', {}, true);
        expect(response.text).toEqual('Component is not running');
      })
      .then(done));

  it('cannot proxy a component', (done) =>
    request(server)
      .get('/proxy/bbc-morph-baz')
      .expect('Content-Type', 'text/html; charset=utf-8')
      .expect(500)
      .then((response: request.Response): void => {
        expect(mockService.request).toHaveBeenCalledWith('bbc-morph-baz', {}, false);
        expect(response.text).toEqual('Component is not running');
      })
      .then(done));
});

describe('given a data component is running', () => {
  beforeEach(() => {
    mockServiceRequest('{ "baz": "123" }', {}, 200);
  });

  it('can be requested without props', (done) =>
    request(server)
      .get('/data/bbc-morph-baz')
      .expect(200)
      .then((response: request.Response) => {
        expect(mockService.request).toHaveBeenCalledWith('bbc-morph-baz', {}, true);
        expect(JSON.parse(response.text)).toEqual({ baz: '123' });
      })
      .then(done));

  it('can be requested with props', (done) =>
    request(server)
      .get('/data/bbc-morph-baz/a/1/b%2Fb/2%202/c/3')
      .expect(200)
      .then((response: request.Response) => {
        expect(mockService.request).toHaveBeenCalledWith(
          'bbc-morph-baz',
          {
            'a': '1',
            'b/b': '2 2',
            'c': '3',
          },
          true,
        );
        expect(JSON.parse(response.text)).toEqual({ baz: '123' });
      })
      .then(done));

  it('correctly handles errors', (done) => {
    (mockService.request as jest.Mock).mockReset();
    (mockService.request as jest.Mock).mockImplementationOnce(() => {
      throw new Error('erhmahgerd');
    });
    return request(server)
      .get('/data/bbc-morph-baz')
      .expect(500)
      .then((response: request.Response) => {
        expect(mockService.request).toHaveBeenCalledWith('bbc-morph-baz', {}, true);
        expect(response.text).toEqual('erhmahgerd');
      })
      .then(done);
  });

  it('can be proxied without props', (done) =>
    request(server)
      .get('/proxy/bbc-morph-baz')
      .expect(200)
      .expect('Content-Type', 'text/html; charset=utf-8')
      .then((response: request.Response) => {
        expect(mockService.request).toHaveBeenCalledWith('bbc-morph-baz', {}, false);
        expect(JSON.parse(response.text)).toEqual({ baz: '123' });
      })
      .then(done));

  it('can be proxied with props', (done) =>
    request(server)
      .get('/proxy/bbc-morph-baz/a/1/b%2Fb/2%202/c/3')
      .expect(200)
      .expect('Content-Type', 'text/html; charset=utf-8')
      .then((response: request.Response) => {
        expect(mockService.request).toHaveBeenCalledWith(
          'bbc-morph-baz',
          {
            'a': '1',
            'b/b': '2 2',
            'c': '3',
          },
          false,
        );
        expect(JSON.parse(response.text)).toEqual({ baz: '123' });
      })
      .then(done));

  it('correctly handles errors when proxied', (done) => {
    (mockService.request as jest.Mock).mockReset();
    (mockService.request as jest.Mock).mockImplementationOnce(() => {
      throw new Error('erhmahgerd');
    });
    return request(server)
      .get('/proxy/bbc-morph-baz')
      .expect(500)
      .then((response: request.Response) => {
        expect(mockService.request).toHaveBeenCalledWith('bbc-morph-baz', {}, false);
        expect(response.text).toEqual('erhmahgerd');
      })
      .then(done);
  });

  it('should correctly handle being requested with an accept header of text/html', (done) =>
    request(server)
      .get('/data/bbc-morph-baz')
      .accept('text/html')
      .expect(200)
      .expect('Content-Type', 'text/html; charset=utf-8')
      .then(() => {
        expect(mockService.request).toHaveBeenCalledWith('bbc-morph-baz', {}, true);
      })
      .then(done));
});

describe('given a data component is running and returning non-200 with custom headers', () => {
  beforeEach(() => {
    mockServiceRequest('{ "baz": "123" }', { custom: 'header' }, 222);
  });

  it('correctly forwards status code and headers when requested', (done) =>
    request(server)
      .get('/data/bbc-morph-baz')
      .expect(222)
      .expect('custom', 'header')
      .then((response: request.Response) => {
        expect(JSON.parse(response.text)).toEqual({ baz: '123' });
      })
      .then(done));

  it('correctly forwards status code and headers when proxied', (done) =>
    request(server)
      .get('/proxy/bbc-morph-baz')
      .expect(222)
      .expect('custom', 'header')
      .then((response: request.Response) => {
        expect(JSON.parse(response.text)).toEqual({ baz: '123' });
      })
      .then(done));
});

describe('given a view component is running', () => {
  beforeEach(() => {
    mockServiceRequest(
      '{ "head": ["<link>head1</link>", "<link>head2</link>"], "bodyInline": "<h1>Hello</h1>", "bodyLast": ["<p>bodylast1</p>", "<p>bodylast2</p>"] }',
      {},
      200,
    );
  });

  it('can be requested without props', (done) =>
    request(server)
      .get('/view/bbc-morph-baz')
      .expect(200)
      .expect('Content-Type', 'text/html; charset=utf-8')
      .then((response: request.Response) => {
        expect(mockService.request).toHaveBeenCalledWith('bbc-morph-baz', {}, true);
        expect(response.text).toContain('<h1>Hello</h1>');
        expect(response.text).toMatchSnapshot();
      })
      .then(done));

  it('can be requested with props', (done) =>
    request(server)
      .get('/view/bbc-morph-baz/hello/world')
      .expect(200)
      .expect('Content-Type', 'text/html; charset=utf-8')
      .then((response: request.Response) => {
        expect(mockService.request).toHaveBeenCalledWith(
          'bbc-morph-baz',
          {
            hello: 'world',
          },
          true,
        );
        expect(response.text).toContain('<h1>Hello</h1>');
        expect(response.text).toMatchSnapshot();
      })
      .then(done));

  it('correctly handles errors', (done) => {
    (mockService.request as jest.Mock).mockReset();
    (mockService.request as jest.Mock).mockImplementationOnce(() => {
      throw new Error('erhmahgerd');
    });
    return request(server)
      .get('/view/bbc-morph-baz')
      .expect(500)
      .then((response: request.Response) => {
        expect(mockService.request).toHaveBeenCalledWith('bbc-morph-baz', {}, true);
        expect(response.text).toEqual('erhmahgerd');
      })
      .then(done);
  });

  it('can be proxied without props', (done) =>
    request(server)
      .get('/proxy/bbc-morph-baz')
      .expect(200)
      .expect('Content-Type', 'text/html; charset=utf-8')
      .then((response: request.Response) => {
        expect(mockService.request).toHaveBeenCalledWith('bbc-morph-baz', {}, false);
        expect(response.text).toContain('<h1>Hello</h1>');
      })
      .then(done));

  it('can be proxied with props', (done) =>
    request(server)
      .get('/proxy/bbc-morph-baz/hello/world')
      .expect(200)
      .expect('Content-Type', 'text/html; charset=utf-8')
      .then((response: request.Response) => {
        expect(mockService.request).toHaveBeenCalledWith(
          'bbc-morph-baz',
          {
            hello: 'world',
          },
          false,
        );
        expect(response.text).toContain('<h1>Hello</h1>');
      })
      .then(done));

  it('correctly handles errors when proxied', (done) => {
    (mockService.request as jest.Mock).mockReset();
    (mockService.request as jest.Mock).mockImplementationOnce(() => {
      throw new Error('erhmahgerd');
    });
    return request(server)
      .get('/proxy/bbc-morph-baz')
      .expect(500)
      .then((response: request.Response) => {
        expect(mockService.request).toHaveBeenCalledWith('bbc-morph-baz', {}, false);
        expect(response.text).toEqual('erhmahgerd');
      })
      .then(done);
  });

  it('should correctly handle being requested with an accept header of text/html', (done) =>
    request(server)
      .get('/view/bbc-morph-baz')
      .accept('text/html')
      .expect(200)
      .expect('Content-Type', 'text/html; charset=utf-8')
      .then(() => {
        expect(mockService.request).toHaveBeenCalledWith('bbc-morph-baz', {}, true);
      })
      .then(done));

  describe('and the component has no head or bodyLast', () => {
    beforeEach(() => {
      (mockService.request as jest.Mock).mockReset();
      mockServiceRequest('{ "bodyInline": "<h1>Hello</h1>" }', {}, 200);
    });

    it('should return the expected values when requested', (done) =>
      request(server)
        .get('/view/bbc-morph-baz')
        .expect(200)
        .expect('Content-Type', 'text/html; charset=utf-8')
        .then((response: request.Response) => {
          expect(mockService.request).toHaveBeenCalledWith('bbc-morph-baz', {}, true);
          expect(response.text).toContain('<h1>Hello</h1>');
          expect(response.text).toMatchSnapshot();
        })
        .then(done));
  });
});

describe('given a view component is running and returning non-200 with custom headers', () => {
  beforeEach(() => {
    mockServiceRequest('non-200 response', { custom: 'header' }, 222);
  });

  it('correctly forwards status code and headers and returns an unwrapped body when requested', (done) =>
    request(server)
      .get('/view/bbc-morph-baz')
      .expect(222)
      .expect('custom', 'header')
      .then((response: request.Response) => {
        expect(response.text).toEqual('non-200 response');
      })
      .then(done));

  it('correctly forwards status code and headers and returns an unwrapped body when proxied', (done) =>
    request(server)
      .get('/proxy/bbc-morph-baz')
      .expect(222)
      .expect('custom', 'header')
      .then((response: request.Response) => {
        expect(response.text).toEqual('non-200 response');
      })
      .then(done));
});

describe('local-push.js', () => {
  describe('when there is a config value set for livePushPollInterval', () => {
    beforeEach(() => {
      mockConfig.get.mockReturnValue('42');
    });

    it('returns the file with the expected poll interval', (done) =>
      request(server)
        .get('/local-push.js')
        .expect(200)
        .expect('Content-Type', 'application/javascript; charset=utf-8')
        .then((response: request.Response) => {
          expect(mockConfig.get).toHaveBeenCalledWith('livePushPollInterval');
          expect(response.text).toContain('intervals[topic] = setInterval(poll, 42);');
          expect(response.text).toMatchSnapshot();
        })
        .then(done));
  });

  describe('when there is no config value set for livePushPollInterval', () => {
    beforeEach(() => {
      mockConfig.get.mockReturnValue(undefined);
    });

    it('returns the file with the default poll interval', (done) =>
      request(server)
        .get('/local-push.js')
        .expect(200)
        .expect('Content-Type', 'application/javascript; charset=utf-8')
        .then((response: request.Response) => {
          expect(mockConfig.get).toHaveBeenCalledWith('livePushPollInterval');
          expect(response.text).toContain('intervals[topic] = setInterval(poll, 10000);');
          expect(response.text).toMatchSnapshot();
        })
        .then(done));
  });
});
