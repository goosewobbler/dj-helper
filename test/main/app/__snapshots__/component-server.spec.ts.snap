// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`given a view component is running and the component has no head or bodyLast should return the expected values when requested 1`] = `
"<!doctype html>
   <html class=\\"b-pw-1280\\">
     <head>
       <meta charset=\\"utf-8\\">
       <meta http-equiv=\\"x-ua-compatible\\" content=\\"ie=edge\\">
       <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1\\">
       <link rel=\\"shortcut icon\\" type=\\"image/png\\" href=\\"http://localhost:undefined/image/icon/morph.png\\"/>
       <script src=\\"http://localhost:undefined/socket.io/socket.io.js\\"></script>
       
       <style>body {font-size: 62.5%;line-height: 1;}</style>
     </head>
     <body>
       <h1>Hello</h1>
       <script src=\\"//m.int.files.bbci.co.uk/modules/vendor/requirejs/2.1.20/require.min.js\\"></script>
       
       <script>const socket = io(\\"http://localhost:undefined\\"); socket.on(\\"reload\\", () => window.location.reload(true));</script>
     </body>
   </html>"
`;

exports[`given a view component is running can be requested with props 1`] = `
"<!doctype html>
   <html class=\\"b-pw-1280\\">
     <head>
       <meta charset=\\"utf-8\\">
       <meta http-equiv=\\"x-ua-compatible\\" content=\\"ie=edge\\">
       <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1\\">
       <link rel=\\"shortcut icon\\" type=\\"image/png\\" href=\\"http://localhost:undefined/image/icon/morph.png\\"/>
       <script src=\\"http://localhost:undefined/socket.io/socket.io.js\\"></script>
       <link>head1</link><link>head2</link>
       <style>body {font-size: 62.5%;line-height: 1;}</style>
     </head>
     <body>
       <h1>Hello</h1>
       <script src=\\"//m.int.files.bbci.co.uk/modules/vendor/requirejs/2.1.20/require.min.js\\"></script>
       <p>bodylast1</p><p>bodylast2</p>
       <script>const socket = io(\\"http://localhost:undefined\\"); socket.on(\\"reload\\", () => window.location.reload(true));</script>
     </body>
   </html>"
`;

exports[`given a view component is running can be requested without props 1`] = `
"<!doctype html>
   <html class=\\"b-pw-1280\\">
     <head>
       <meta charset=\\"utf-8\\">
       <meta http-equiv=\\"x-ua-compatible\\" content=\\"ie=edge\\">
       <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1\\">
       <link rel=\\"shortcut icon\\" type=\\"image/png\\" href=\\"http://localhost:undefined/image/icon/morph.png\\"/>
       <script src=\\"http://localhost:undefined/socket.io/socket.io.js\\"></script>
       <link>head1</link><link>head2</link>
       <style>body {font-size: 62.5%;line-height: 1;}</style>
     </head>
     <body>
       <h1>Hello</h1>
       <script src=\\"//m.int.files.bbci.co.uk/modules/vendor/requirejs/2.1.20/require.min.js\\"></script>
       <p>bodylast1</p><p>bodylast2</p>
       <script>const socket = io(\\"http://localhost:undefined\\"); socket.on(\\"reload\\", () => window.location.reload(true));</script>
     </body>
   </html>"
`;

exports[`local-push.js when there is a config value set for livePushPollInterval returns the file with the expected poll interval 1`] = `
"/* global window, XMLHttpRequest */
/* eslint-disable no-console */
window.define('live-push', [], () => {
  const intervals = {};
  const onMomentCallbacks = [];
  const consoleAvailable = !!window.console && !window.jasmine;
  const exports = {};

  function logSubscriptions() {
    console.info('Morph Live Push Local: subscriptions:', Object.keys(intervals));
  }

  function callbackAllOnMomentCallbacks(moment) {
    onMomentCallbacks.forEach((callback) => {
      callback(moment);
    });
  }

  function fetch(topic, callbackWhen202, callback) {
    let url = \`http://localhost:4000/proxy\${topic.substring(12)}\`;

    if (topic.indexOf('nitro://') > -1) {
      url = \`http://open.test.bbci.co.uk/live-broker/moments?request=\${encodeURIComponent(topic)}\`;
    }

    const xhr = new XMLHttpRequest();
    xhr.open('GET', url, true);
    xhr.onreadystatechange = () => {
      let responseBody;
      if (xhr.readyState === 4) {
        const responseCode = xhr.status;
        if (responseCode === 200) {
          responseBody = xhr.responseText;
          if (topic.indexOf('nitro://') > -1) {
            // The LiveBroker returns a different format to Morph, so just return the raw moments
            responseBody = JSON.parse(xhr.responseText).moments;
          }
          callback({
            topic,
            payload: responseBody,
          });
        } else if (consoleAvailable) {
          console.warn('Morph Live Push Local: response code:', responseCode, 'for:', url);
          if (responseCode === 202 && callbackWhen202) {
            callback();
          }
        }
      }
    };
    xhr.send(null);
  }

  exports.subscribe = (topic) => {
    if (intervals[topic]) {
      return;
    }

    const poll = fetch.bind(this, topic, false, callbackAllOnMomentCallbacks);
    intervals[topic] = setInterval(poll, 42); //eslint-disable-line
    poll();

    if (consoleAvailable) {
      console.info('Morph Live Push Local: subscribed to:', topic);
      logSubscriptions();
    }
  };

  exports.unsubscribe = (topic) => {
    if (intervals[topic]) {
      clearInterval(intervals[topic]);
      delete intervals[topic];
      if (consoleAvailable) {
        console.info('Morph Live Push Local: unsubscribed from:', topic);
        logSubscriptions();
      }
    }
  };

  exports.start = () => {};

  exports.get = (topic, count, callback) => {
    if (consoleAvailable) {
      console.info('Morph Live Push Local: get (XHR):', topic);
    }

    fetch(topic, true, (moment) => {
      const moments = moment ? [moment] : [];
      callback(JSON.stringify({ moments }));
    });
  };

  exports.on = (channel, callback) => {
    onMomentCallbacks.push(callback);
  };

  exports.getInstance = () => {
    if (consoleAvailable) {
      console.info('Morph Live Push Local: getting instance...');
    }
    return exports;
  };

  return exports;
});
"
`;

exports[`local-push.js when there is no config value set for livePushPollInterval returns the file with the default poll interval 1`] = `
"/* global window, XMLHttpRequest */
/* eslint-disable no-console */
window.define('live-push', [], () => {
  const intervals = {};
  const onMomentCallbacks = [];
  const consoleAvailable = !!window.console && !window.jasmine;
  const exports = {};

  function logSubscriptions() {
    console.info('Morph Live Push Local: subscriptions:', Object.keys(intervals));
  }

  function callbackAllOnMomentCallbacks(moment) {
    onMomentCallbacks.forEach((callback) => {
      callback(moment);
    });
  }

  function fetch(topic, callbackWhen202, callback) {
    let url = \`http://localhost:4000/proxy\${topic.substring(12)}\`;

    if (topic.indexOf('nitro://') > -1) {
      url = \`http://open.test.bbci.co.uk/live-broker/moments?request=\${encodeURIComponent(topic)}\`;
    }

    const xhr = new XMLHttpRequest();
    xhr.open('GET', url, true);
    xhr.onreadystatechange = () => {
      let responseBody;
      if (xhr.readyState === 4) {
        const responseCode = xhr.status;
        if (responseCode === 200) {
          responseBody = xhr.responseText;
          if (topic.indexOf('nitro://') > -1) {
            // The LiveBroker returns a different format to Morph, so just return the raw moments
            responseBody = JSON.parse(xhr.responseText).moments;
          }
          callback({
            topic,
            payload: responseBody,
          });
        } else if (consoleAvailable) {
          console.warn('Morph Live Push Local: response code:', responseCode, 'for:', url);
          if (responseCode === 202 && callbackWhen202) {
            callback();
          }
        }
      }
    };
    xhr.send(null);
  }

  exports.subscribe = (topic) => {
    if (intervals[topic]) {
      return;
    }

    const poll = fetch.bind(this, topic, false, callbackAllOnMomentCallbacks);
    intervals[topic] = setInterval(poll, 10000); //eslint-disable-line
    poll();

    if (consoleAvailable) {
      console.info('Morph Live Push Local: subscribed to:', topic);
      logSubscriptions();
    }
  };

  exports.unsubscribe = (topic) => {
    if (intervals[topic]) {
      clearInterval(intervals[topic]);
      delete intervals[topic];
      if (consoleAvailable) {
        console.info('Morph Live Push Local: unsubscribed from:', topic);
        logSubscriptions();
      }
    }
  };

  exports.start = () => {};

  exports.get = (topic, count, callback) => {
    if (consoleAvailable) {
      console.info('Morph Live Push Local: get (XHR):', topic);
    }

    fetch(topic, true, (moment) => {
      const moments = moment ? [moment] : [];
      callback(JSON.stringify({ moments }));
    });
  };

  exports.on = (channel, callback) => {
    onMomentCallbacks.push(callback);
  };

  exports.getInstance = () => {
    if (consoleAvailable) {
      console.info('Morph Live Push Local: getting instance...');
    }
    return exports;
  };

  return exports;
});
"
`;
